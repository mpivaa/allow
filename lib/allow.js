'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.config = undefined;

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.setupAllow = setupAllow;
exports.isAllowed = isAllowed;

var _lodash = require('lodash');

function _toArray(arr) { return Array.isArray(arr) ? arr : Array.from(arr); }

var config = exports.config = {
  levels: ['app'],
  default: 'app',
  aliases: {
    'a': 'app'
  }
};

function setupAllow(newConfig) {
  exports.config = config = _extends({}, config, newConfig);
}

function isAllowed(ctx, allowedRoles) {
  if (isValidContext(ctx)) {
    return (0, _lodash.some)(parseRoles(allowedRoles), function (_ref) {
      var _ref2 = _slicedToArray(_ref, 2),
          level = _ref2[0],
          role = _ref2[1];

      if (isValidContext(ctx)) {
        return checkRole([level, role], ctx);
      }
      return false;
    });
  } else {
    return false;
  }
}

function checkRole(_ref3, ctx) {
  var _ref4 = _slicedToArray(_ref3, 2),
      level = _ref4[0],
      role = _ref4[1];

  var roleOnLevel = getRoleOnLevel(level, ctx);
  if (!(0, _lodash.isNil)(roleOnLevel)) {
    return role === "*" || roleOnLevel === role;
  } else {
    return false;
  }
}

function parseRoles(allowedRoles) {
  return (0, _lodash.map)(allowedRoles, function (composedRole) {
    var _composedRole$split = composedRole.split(':'),
        _composedRole$split2 = _toArray(_composedRole$split),
        level = _composedRole$split2[0],
        role = _composedRole$split2[1],
        ignore = _composedRole$split2.slice(2);

    var expandedLevel = expandLevel(level);
    return [expandedLevel, role];
  });
}

function getRoleOnLevel(level, ctx) {
  var user = ctx.user;

  if (level === config.default) {
    return (0, _lodash.get)(user, ['roles', level], null);
  } else {
    var id = (0, _lodash.get)(ctx, [level, 'id'], null);
    return (0, _lodash.get)(user, ['roles', level, id], null);
  }
}

function expandLevel(level) {
  if (!(0, _lodash.isNil)(config.aliases[level])) {
    return config.aliases[level];
  } else {
    return level;
  }
}

function isValidContext(ctx) {
  return !(0, _lodash.isEmpty)(ctx) && !(0, _lodash.isNil)(ctx) && !(0, _lodash.isNil)(ctx.user);
}