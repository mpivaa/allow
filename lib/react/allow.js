'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Allow = Allow;

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _allow = require('../allow');

var _context = require('./context');

var _context2 = _interopRequireDefault(_context);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function Allow(_ref) {
  var children = _ref.children,
      roles = _ref.roles,
      IfNotAllowed = _ref.ifNotAllowed,
      overrideContext = _ref.overrideContext;


  return React.createElement(
    _context2.default.Consumer,
    null,
    function (_ref2) {
      var appContext = _ref2.context;

      var ctx = overrideContext || appContext;
      if ((0, _allow.isAllowed)(ctx, roles)) {
        return children;
      } else if (IfNotAllowed) {
        return React.createElement(IfNotAllowed, null);
      } else {
        return null;
      }
    }
  );
}