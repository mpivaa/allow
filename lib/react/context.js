'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Provider = exports.Consumer = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _lodash = require('lodash');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var initialValues = {
  context: null
};

var ReactContext = React.createContext(initialValues);
var Consumer = ReactContext.Consumer;

var Provider = function (_React$Component) {
  _inherits(Provider, _React$Component);

  function Provider() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Provider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Provider.__proto__ || Object.getPrototypeOf(Provider)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      context: _this.props.defaultContext
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Provider, [{
    key: 'getContext',
    value: function getContext(parentContext) {
      if (this.isControlled()) {
        if (parentContext) {
          return _extends({}, parentContext, this.props.context);
        } else {
          return this.props.context;
        }
      } else {
        return this.state.context;
      }
    }
  }, {
    key: 'isControlled',
    value: function isControlled() {
      return true || !(0, _lodash.isNil)(this.props.context);
    }
  }, {
    key: 'getProviderValue',
    value: function getProviderValue(parentContext) {
      var _this2 = this;

      return {
        context: this.getContext(parentContext),
        updateContext: function updateContext(context) {
          if (!_this2.isControlled()) {
            _this2.setState({ context: context });
          }
          if ((0, _lodash.isFunction)(_this2.props.onContextChange)) {
            _this2.props.onContextChange(context);
          }
        }
      };
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      return React.createElement(
        ReactContext.Consumer,
        null,
        function (_ref2) {
          var context = _ref2.context;

          return React.createElement(
            ReactContext.Provider,
            { value: _this3.getProviderValue(context) },
            _this3.props.children
          );
        }
      );
    }
  }]);

  return Provider;
}(React.Component);

exports.Consumer = Consumer;
exports.Provider = Provider;
exports.default = { Consumer: Consumer, Provider: Provider };