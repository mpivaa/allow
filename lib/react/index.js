'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _context = require('./context');

Object.defineProperty(exports, 'AllowContext', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_context).default;
  }
});

var _allow = require('./allow');

Object.defineProperty(exports, 'Allow', {
  enumerable: true,
  get: function get() {
    return _allow.Allow;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }