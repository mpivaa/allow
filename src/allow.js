// @flow
import { isEmpty, some, map, slice, isNil, isUndefined, get } from 'lodash'
import type {
  Context,
  Level,
  Role,
  ComposedRole,
  Config
} from './types';

export let config: Config = {
  levels: ['app'],
  default: 'app',
  aliases: {
    'a': 'app'
  }
}

export function setupAllow(newConfig: Config) {
  config = {...config, ...newConfig}
}

export function isAllowed(ctx: ?Context, allowedRoles: [ComposedRole]): boolean {
  if(isValidContext(ctx)) {
    return some(parseRoles(allowedRoles), ([level, role]) => {
      if(isValidContext(ctx)) {
        return checkRole([level, role], ctx)
      }
      return false
    })
  } else {
    return false
  }
}

function checkRole([level: Level, role: Role], ctx: Context): boolean {
  const roleOnLevel = getRoleOnLevel(level, ctx)
  if(!isNil(roleOnLevel)) {
    return role === "*" || roleOnLevel === role
  } else {
    return false
  }
}

function parseRoles(allowedRoles: [ComposedRole]): Array<[Level, Role]> {
  return map(allowedRoles, composedRole => {
    const [level, role, ...ignore] = composedRole.split(':')
    const expandedLevel = expandLevel(level)
    return [expandedLevel, role]
  })
}

function getRoleOnLevel(level: Level, ctx: Context): ?Role {
  const { user } = ctx
  if(level === config.default) {
    return get(user, ['roles', level], null)
  } else {
    const id = get(ctx, [level, 'id'], null)
    return get(user, ['roles', level, id], null)
  }
}

function expandLevel(level: Level) {
  if(!isNil(config.aliases[level])) {
    return config.aliases[level]
  } else {
    return level
  }
}

function isValidContext(ctx: ?Context): boolean %checks {
  return !isEmpty(ctx) && !isNil(ctx) && !isNil(ctx.user)
}