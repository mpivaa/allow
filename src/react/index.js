export { default as AllowContext } from './context'
export { Allow } from './auth'