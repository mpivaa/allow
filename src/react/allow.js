// @flow
import * as React from 'react'
import type { Context, ComposedRole } from '../types'
import { isAllowed } from '../allow'
import AllowContext from './context'

type Props = {
  children?: React.Node,
  roles?: [ComposedRole],
  ifNotAllowed?: React.ComponentType<any>,
  overrideContext?: Context
}

type ConsumerValue = {
  context: ?Context
}

export function Allow({
  children,
  roles,
  ifNotAllowed: IfNotAllowed,
  overrideContext
}: Props) {

  return (
    <AllowContext.Consumer>
      {({ context: appContext }: ConsumerValue): React.Node => {
        const ctx = overrideContext || appContext
        if(isAllowed(ctx, roles)) {
          return children
        } else if(IfNotAllowed) {
          return <IfNotAllowed />
        } else {
          return null
        }
      }}
    </AllowContext.Consumer>
  )
}