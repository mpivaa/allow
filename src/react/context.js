// @flow
import * as React from 'react'
import type { Context } from '../types'
import { isFunction, isNil } from 'lodash'

type InitialValues = {
  context: ?Context
}

type Props = {
  children?: React.Node,
  context?: Context,
  defaultContext?: Context,
  onContextChange?: (context: ?Context) => void
}

type State = {
  context: ?Context
}

const initialValues = ({
  context: null
}: InitialValues)

const ReactContext  = React.createContext(initialValues)
const { Consumer } = ReactContext


class Provider extends React.Component<Props, State> {
  state = {
    context: this.props.defaultContext
  }

  getContext(parentContext: ?Context) {
    if(this.isControlled()) {
      if(parentContext) {
        return {
          ...parentContext,
          ...this.props.context
        }
      } else {
        return this.props.context
      }
    } else {
      return this.state.context
    }
  }

  isControlled() {
    // TODO: Remove uncontrolled options
    return true || !isNil(this.props.context)
  }

  getProviderValue(parentContext: ?Context) {
    return {
      context: this.getContext(parentContext),
      // TODO: updateContext
      // updateContext: (context: ?Context): void => {
      //   if(!this.isControlled()) {
      //     this.setState({ context })
      //   }
      //   if(isFunction(this.props.onContextChange)) {
      //     this.props.onContextChange(context)
      //   }
      // }
    }
  }

  render() {
    return (
      <ReactContext.Consumer>
        {({ context }): React.Node => {
          return (
            <ReactContext.Provider value={this.getProviderValue(context)}>
              {this.props.children}
            </ReactContext.Provider>
          )
        }}
      </ReactContext.Consumer>
    )
  }
}

export { Consumer, Provider }
export default { Consumer, Provider }
