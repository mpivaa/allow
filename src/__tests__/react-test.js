import React from 'react'
import AllowContext from '../react/context'
import { Allow } from '../react/allow'
import { setupAllow } from '../allow'
import renderer from 'react-test-renderer'

setupAllow({
  levels: ['app', 'organization'],
  default: 'app',
  aliases: {
    'a': 'app'
  }
})

const ctx = {
  user: {
    roles: {
      app: 'admin',
      organization: { '1': 'admin' }
    }
  }
}

test('Render children when is allowed', () => {
  const component = renderer.create(
    <Allow roles={['app:admin']} overrideContext={ctx}>
      This should appear
    </Allow>
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Render nothing when is not allowed', () => {
  const component = renderer.create(
    <Allow roles={['app:somerole']} overrideContext={ctx}>
      This should not appear
    </Allow>
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Use context to allow', () => {
  const component = renderer.create(
    <AllowContext.Provider context={ctx}>
      <Allow roles={['app:admin']}>
        This should appear
      </Allow>
    </AllowContext.Provider>
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Render ifNotAllowed', () => {
  const component = renderer.create(
    <AllowContext.Provider context={ctx}>
      <Allow roles={['app:somerole']} ifNotAllowed={(props) => '404'}>
        This should not appear
      </Allow>
    </AllowContext.Provider>
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Nested provider should override parent', () => {
  const organization = { id: 1 }
  const component = renderer.create(
    <AllowContext.Provider context={ctx}>
      <AllowContext.Provider context={{ organization }}>
        <Allow roles={['organization:admin']}>
          This should appear
        </Allow>
      </AllowContext.Provider>
    </AllowContext.Provider>
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Nested provider should not affect outer components', () => {
  const organization = { id: 1 }
  const component = renderer.create(
    <AllowContext.Provider context={ctx}>
      <AllowContext.Provider context={{ organization }}>
        <Allow roles={['organization:admin']}>
          This should appear
        </Allow>
      </AllowContext.Provider>

      <Allow roles={['organization:admin']}>
        This should not appear
      </Allow>
    </AllowContext.Provider>
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})