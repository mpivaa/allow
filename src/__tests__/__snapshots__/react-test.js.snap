// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nested provider should not affect outer components 1`] = `"This should appear"`;

exports[`Nested provider should override parent 1`] = `"This should appear"`;

exports[`Render children when is allowed 1`] = `"This should appear"`;

exports[`Render ifNotAllowed 1`] = `"404"`;

exports[`Render nothing when is not allowed 1`] = `null`;

exports[`Use context to allow 1`] = `"This should appear"`;
